### PR TITLE
[devutils] Enhance devutils

### PR DESCRIPTION
**Pull request by bingwang-ms**
Signed-off-by: bingwang <bingwang@microsoft.com>

### Description of PR

Summary:
This PR enhances ```devutils```
1. Support automatic login
2. Support running ssh command on different hosts concurrently
3. Beautiful output

### Type of change

- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

#### What is the motivation for this PR?
This PR is to enhance ```devutils``` tool.

#### How did you do it?
Please refer to code.

#### How did you verify/test it?
All actions are verified.

#### Any platform specific information?
No.

#### Supported testbed topology if it's a new test case?
No.

### CODE REVIEW

File: ansible/devutil/ssh_utils.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+import paramiko
+from paramiko.py3compat import u
+import termios
+import tty
+import select
+import sys
+import socket
+
+class SSHClient(paramiko.client.SSHClient):
+
```
Comment (yxieca) [inline ansible/devutil/ssh_utils.py:10]: Remove extra blank line?

File: ansible/devutil/ssh_utils.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            tty.setraw(sys.stdin.fileno())
+            tty.setcbreak(sys.stdin.fileno())
+            chan = super(SSHClient, self).invoke_shell()
+            chan.settimeout(0.0)
+
+            while True:
+                r, w, e = select.select([chan, sys.stdin], [], [])
+                if chan in r:
+                    try:
+                        x = u(chan.recv(1024))
```
Comment (yxieca) [inline ansible/devutil/ssh_utils.py:44]: Maybe define a max line constant? Is 1024 enough?

File: ansible/devutils [hunk 3]
```diff
@@ -18,80 +30,136 @@ def run_cmd(cmd):
...
+                                    console_port=vars['console_port'],
+                                    sonic_username=vars['creds']['username'],
+                                    sonic_password=vars['creds']['password'],
+                                    console_username=vars['creds']['console_user'][vars['console_type']],
+                                    console_password=vars['creds']['console_password'][vars['console_type']])
+        console_host.posix_shell()
+
+def action_dispatcher(parameters):
+    # Actions that can run simultaneously in different hosts
+    parallel_actions = ['ssh_run_command']
```
Comment (yxieca) [inline ansible/devutils:87]: Can ping also be executed in parallel?

File: ansible/devutils [hunk 4]
```diff
@@ -18,80 +30,136 @@ def run_cmd(cmd):
...
+                        type=int, required=False, default=1)
 
     args = parser.parse_args()
-
-    hosts = get_host_list(args.inventory, args.category)
+    if not validate_args(args):
+        return
+    global g_task_runner, g_inv_mgr
+    g_task_runner = TaskRunner(max_worker=args.concurrency)
+    g_inv_mgr = HostManager(args.inventory)
```
Comment (yxieca) [inline ansible/devutils:147]: Lack handling code when no ansible?

File: ansible/devutil/inv_helpers.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        hosts = {}
+        for hostname, _ in self._inv_mgr.hosts.items():
+            hosts.update({hostname: self.get_host_vars(hostname)})
+        return hosts
+
+    def get_host_list(self, category, limit=None):
+        if not limit or limit == '':
+            limit = '*'
+        res = {}
+        hosts = self._inv_mgr.get_hosts(pattern=limit)
```
Comment (yxieca) [inline ansible/devutil/inv_helpers.py:69]: Can we also add support to match limit with testbed name here?

**Review by yxieca — COMMENTED** [review]
Great enhancement! Thanks Bing!

File: ansible/devutil/ssh_utils.py (hunk 1 above)
Comment (bingwang-ms) [reply ansible/devutil/ssh_utils.py]: Thanks. Updated.

File: ansible/devutil/ssh_utils.py (hunk 2 above)
Comment (bingwang-ms) [reply ansible/devutil/ssh_utils.py:62]: Yes. I think 1024 is enough. Actually, if the output is more than 1024 bytes, the ```select``` in loop will be triggered again, then we have another opportunity to read.

File: ansible/devutils (hunk 4 above)
Comment (bingwang-ms) [reply ansible/devutils:147]: Yes. The code without ```Ansible``` is not implemented yet. I added a todo here.

File: ansible/devutil/inv_helpers.py (hunk 5 above)
Comment (bingwang-ms) [reply ansible/devutil/inv_helpers.py:85]: Good suggestion. Added to todo list. Thanks

**Review by yxieca — APPROVED** [review]

**Pull request merged**